### PR TITLE
WIP: Runtime filters gRPC definitions

### PR DESCRIPTION
**Pull request by Molter73**
## Description

WIP, draft for gRPC communication of runtime filters feature.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

TODO(replace-me)  
Use this space to explain **how you validated** that **your change functions exactly how you expect it**.
Feel free to attach JSON snippets, curl commands, screenshots, etc. Apply a simple benchmark: would the information you
provided convince any reviewer or any external reader that you did enough to validate your change.

It is acceptable to assume trust and keep this section light, e.g. as a bullet-point list.

It is acceptable to skip testing in cases when CI is sufficient, or it's a markdown or code comment change only.  
It is also acceptable to skip testing for changes that are too taxing to test before merging. In such case you are
responsible for the change after it gets merged which includes reverting, fixing, etc. Make sure you validate the change
ASAP after it gets merged or explain in PR when the validation will be performed.  
Explain here why you skipped testing in case you did so.

Have you created automated tests for your change? Explain here which validation activities you did manually and why so.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Comment by JoukoVirtanen** [comment]
Looks good. There should be something like a state field. We need to specify what impact this is having. Are we turning the feature on or off? Maybe we can specify that a feature is on for a cluster and then turn it off for specific namespaces, or specify that a feature is off for a cluster and then turn it on for specific namespaces.

File: proto/storage/runtime_filters.proto [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  enum RuntimeFiltersType {
+    STRING = 0; // Exact match by comparing namespaces to the string directly.
+    REGEX = 1;  // Regex applied to namespaces in search of a match.
+  }
+
+  enum RuntimeFiltersEntity {
+    NAMESPACE = 0;
+    DEPLOYMENT = 1;
```
Comment (JoukoVirtanen) [inline proto/storage/runtime_filters.proto:21]:
```suggestion
    DEPLOYMENT = 1;
    CLUSTER = 2;
```

File: proto/storage/runtime_filters.proto [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  RuntimeFilterFeatures feature = 1;
+  string pattern = 2;
+  RuntimeFiltersType type = 3;
+  RuntimeFiltersEntity entity = 4;
```
Comment (JoukoVirtanen) [inline proto/storage/runtime_filters.proto:27]:
```suggestion
  RuntimeFiltersEntity entity = 4;
  boolean enabled;
```

File: proto/storage/runtime_filters.proto (hunk 1 above)
Comment (JoukoVirtanen) [reply proto/storage/runtime_filters.proto:21]: We should be able to turn features on and off at the cluster level as well.

Comment (JoukoVirtanen) [reply proto/storage/runtime_filters.proto:21]: The customer should be able to turn features on and off at the cluster level as well.

File: proto/storage/runtime_filters.proto (hunk 2 above)
Comment (JoukoVirtanen) [reply proto/storage/runtime_filters.proto:27]: The action field can be used to specify if a field is being enabled, disabled, or something else.

**Comment by Molter73** [comment]
> There should be something like an action field, which specifies if a feature is on or off, or if something else should be done. We need to specify what impact this is having. Are we turning the feature on or off? Maybe we can specify that a feature is on for a cluster and then turn it off for specific namespaces, or specify that a feature is off for a cluster and then turn it on for specific namespaces.

Initially I thought of using an exclamation point in front of the pattern used for configuration as a way to mean this is a negative filter (meaning it will reject things it matches), but I think an additional field makes more sense. Regardless, we should be careful of complexity, it sounds great to be able to turn off a feature at a cluster level and then re-enable it at a namespace level, but if we are also thinking of implementing regexes/globs for the pattern, which implies the possibility of nested matches that turn things on and off, the amount of combinations we need to test will sky-rocket. I think I would rather us have a simple first implementation on the original ask (which was filtering at the namespace level) and slowly scale it from there, I added deployment names more as a placeholder rather than anything else.

**Pull request closed without merging**